In [1]:
import pandas as pd
import re

pd.set_option('display.max_rows', None)
pd.set_option('max_colwidth', 120)

df = pd.read_excel('test_SM.xlsx', sheet_name='Массив')
df.columns = ['Наименование', 'Комментарий', 'Продуктовая группа', 'Группа', 'Производитель', 'ед изм']

## Поиск существительных
### Подготовка

In [2]:
# перевод всех буквенных символов в нижний регистр
names = df['Наименование'].tolist()

for i in range(len(names)):
    names[i] = names[i].lower()

In [3]:
# замена группы символов на один
def replace_all(simb,text, subs):
    return re.sub(simb, subs, text)

In [4]:
for i in range(len(names)):
    names[i] = replace_all('[-)/,(№1234567890]', names[i], ' ')

In [5]:
# встака пробела после точек и запятых
for i in range(len(names)):
    names[i] = names[i].replace('.', '. ', names[i].count('.')).replace(',', ', ', names[i].count(','))

In [6]:
# замена множества пробелов на один пробел
def clean_spaces(text):
    return re.sub(r'\s+', ' ', text)

In [7]:
for i in range(len(names)):
    names[i] = clean_spaces(names[i])

### Создание первого варианта словаря существительных
Для работы использован словарь существительных из 51 тыс. слов по словарю Ефремовой, составленый в рамках проекта Russian Nouns by Harrix

In [8]:
f = open('n_dict.txt', 'r', encoding='utf8')
dict_noun = f.read().splitlines()
f.close()

### Поиск и замена сокращений
В данном случае сокращением будет считаться последовательность буквенных символов длинной 1 или более с одной точкой в конце. Расшифровка сокращений будет присвоена вручную, сокращения, значение которых не удалось определить будут оставлены без изменений, сокращения не несущие смысла или ошибочно попавшие в список будут заменениы на пробел

In [9]:
shorts = []
pos = [] # список содержит порядковые номера строк для дальнейшего просмотра при определении значения сокращения

pattern = r'\w{1,}\.'

for i in range(len(names)):
    short_name = re.findall(pattern, names[i])
    if short_name:
        pos.append(i)
        if short_name not in shorts:
            shorts.append(short_name)

shorts_c = [item for sublist in shorts for item in sublist]

In [10]:
# просмотр строк с сокращениями
print(f"строк с сокращениями - {len(pos)} , сокращений - {len(shorts_c)}")

#for p in pos:
    #print(names[p])

строк с сокращениями - 139 , сокращений - 92


In [11]:
# замена сокращений

# назначаем перевод сокращениий вручную
shorts_mean = ['вспомогательный ', 'передний ', 'ходовой ', 'клапан ', 'ячейка', 'нижний', 'эксцентрическая ', 'уплотнитель ',
               'шестигранный ', 'гидравлический ', 'штука ', 'электро ', 'кольцевое ',
               ' ', 'поставляется ', ' ', 'упаковка ', 'штука ', ' ', ' ', 'анкерный ', 'пластина ',
               'верхний ', 'измерительный ', 'класс ', 'колосниковый ', 'анкерный ', 'пластина ', 'нижний', 'установочный ',
               'колосниковый ', 'гранной ', 'оцинкованый ', 'резиновый ', 'футеровочная ', 'предохранительный ',
               'комплект ', 'элемент ', 'индикатор ', 'индикатор ', ' ', 'гидравлический ', 'конвеер ',
               'неприрывная ', 'промежуточная ', 'ответной ', 'уплотнение ', 'регулировочные ', 'сборе ',
               'гидравлический ', ' ', 'направляющая ', 'шланг ', 'брызгозащитное ',
               'ограждение ', 'комплект ', 'колосника ', ' ', 'стальной ', 'системный ', 'чп ', 'промышленный ', 'порт ',
               'использовать ', 'оцинкованный ', ' ', ' ', 'чертежу ', 'мм ', 'пневмопривода ', 'фланец ', ' ',
               'передаточное ', 'отношение ', ' ', ' ', 'град ', 'кв ', 'торцевой ', 'уплотнитель ', ' ',
               'антиконденсатор ', 'ацетатный ', 'шестигранный ', 'головка ', 'оцинкованный ', 'класс ', 'прочности ',
               'тх ', 'электро ', 'взрыво ', 'защитный ']

dict_shorts = dict(zip(shorts_c, shorts_mean))

for i in range(len(names)):
    line = names[i].split()
    for j in line:
        for word in dict_shorts:
            if j.startswith(word):
                names[i] = names[i].replace(word, dict_shorts[word])

### Выделение существительных 
Выделение существительных из строк будет проводится в 2 этапа, так как исходный словарь может не содержать в себе специализированных технических терминов.
На первом этапе будут определены существительные, которые есть в словаре. После этого в ручную будут добавлены существительные из строк, в которых не определилось ни одного существительного.
На втором этапе будут выявлены недостающие существительные.
В строках с наименованием на латинице существительные не будут определены

In [12]:
# выделение существительных - (1 этап)
temp_list = []
noun_list=[]
for i in range(len(names)):
    line = names[i].split()
    for word in line:
        if word in dict_noun:
            temp_list.append(word)
        elif word[:-1] in dict_noun:
            temp_list.append(word)
        elif word[:-2] in dict_noun:
            temp_list.append(word)
    noun_list.append(temp_list)
    temp_list=[]

nouns = pd.Series(noun_list)
df['существительные'] = nouns.values

df.sort_values(by='существительные')

Наименование  \
847                                                                           TRELLEX ABC-EASY 25-130-2000   
134                                                                                              SLIP RING   
360                                                                                Trellex PP 100-500-1500   
138                                                                                            FLANGE PIPE   
359                                                                                Trellex PP  75-750-1500   
964                                                                                 ML-WASHPROTECTION 0,3L   
356                                                                                Trellex PP  75-600-1000   
699                                                                                           THREADED ROD   
355                                                                                 Trellex PP 75-500-1500   
354                                                                                 Trellex PP 75-500-1000   
353                                                                                 Trellex PP 75-500-1000   
153                                                                                             SLIDE RING   
352                                                                                Trellex PP  50-600-1500   
156                                                                       Круглая стальная щетка HIT-RB 28   
695                                                                                             WEAR PLATE   
961                                                               Шинный штекер D-SUB - SUBCON-PLUS-M/AX 9   
351                                                                                Trellex PP  50-500-1500   
163                                                                                               STRAINER   
350                                                                               Trellex PE 40x150x1000 F   
348                                                                           Trellex PP-XL 10+3-1500-3000   
690                                                                                             WEAR PLATE   
531                                                                                         SUPPORT HOLDER   
363                                                                                 Trellex SP 50-500-1500   
689                                                                                             WEAR PLATE   
364                                                                                 Trellex SP 75-500-1500   
127                                                                                                 PULLER   
102                                                                                 M20 botl protector set   
104                       Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp   
105                 Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
106                 Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
107                 Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
369                                                                                 Trellex PT 75-750-1500   
109                       Injection moulded PU trommel panel 305x305x36mm Thick/16x16mm apertures, Hi-temp   
110   Injection moulded PU trommel panel 305x305x36mm Thick/16x16mm apertures, Hi-temp/150mm High Scroll A   
111   Injection moulded PU trommel panel 305x305x36mm Thick/16x16mm apertures, Hi-temp/150mm High Scroll B   
112   Injection moulded PU trommel panel 305x305x36mm Thick/16x16mm apertures, Hi-temp/150mm High weir bar   
792                     

In [13]:
# добавление новых слов
nouns = pd.Series(noun_list)
#nam = pd.Series(names)
df['существительные'] = nouns.values
#df['проверка'] = nam.values

df.sort_values(by='существительные')

new_words = ['картридж', 'дампер', 'резинокерамика', 'подситник', 'проставка', 'штекер', 'гидроцилиндр',
             'лубрикатор', 'клеммник', 'переходник', 'расходомер', 'комплектующие', 'ноутбук', 'корщетка',
             'муфта', 'удлинитель', 'крыльчатка', 'лопаст', 'колосник', 'щетка','комплект', 'индикатор', 'консол',
            'уплотнение', 'уплотнитель', 'гидроцилидр' 'маслопровод', 'решетка', 'позиционер', 'планка', 'пластин',
            'гровер', 'разделитель', 'байпас']

In [14]:
# создание нового словаря с дополнениями
dict_noun.extend(new_words)

In [15]:
# выделение существительных - (2 этап)
temp_list = []
noun_list=[]
for i in range(len(names)):
    line = names[i].split()
    for word in line:
        if word in dict_noun:
            temp_list.append(word)
        elif word[:-1] in dict_noun:
            temp_list.append(word)
        elif word[:-2] in dict_noun:
            temp_list.append(word)
    noun_list.append(temp_list)
    temp_list=[]

In [16]:
# проверка результатов 

nouns = pd.Series(noun_list)
df['существительные'] = nouns.values
df['существительные'] = df['существительные'].astype(str).str[1:-1]

df

Наименование  \
0                                                     Фильтрующий элемент - карманный, тип AJN Long, PN-12   
1                                                            DIN-рейка перфорированная OMEGA 3F, 35х7,5 мм   
2                                                                                     Гайка 20-UNF SQ, GR8   
3                                                                                      Мембрана, тип А 3/4   
4                                                                                        Светодиодная фара   
5                                                        DIN 125 A / ISO 7089 Шайба плоская M16 (Ø17) Цинк   
6                                                                DIN 127 B Шайба пружинная Гровер M16 Цинк   
7                                                                DIN 127 B Шайба пружинная Гровер M20 Цинк   
8                                                                DIN 127 B Шайба пружинная Гровер M24 Цинк   
9                                                                              ШКАЛА АМПЕРМЕТРА 72Х72,200А   
10                                                                 Датчик температуры  (ПартноMер ELETELE)   
11                                                              Уплотнитель кабельных проходов УКПт-140/42   
12                                                                                              ASHER HARD   
13                                                                          ВСПОМОГ. КОНТАКТ OF/SD/SDE/SDV   
14                                    Кабельный ввод М 20 латунный УТ1,5 IP66/IP67/IP68 (d кабеля 6-12 мм)   
15                                 Кабельный ввод М 88х2 латунный УТ1,5 IP66/IP67/IP68 (d кабеля 65-70 мм)   
16                                                         ВЫКЛЮЧАТЕЛЬ NS1000 N 3P+ MICROLOGIC 2.0 В СБОРЕ   
17                                                          3 РАЗДЕЛИТЕЛЯ ПОЛЮСОВ ДЛЯ ПЕРЕДН.ПРИСОЕДИНЕНИЯ   
18                                                                     Удлинительный шланг HIT-VL 16 (10m)   
19                                                       DIN 471 Кольцо стопорное наружное Ø70 x 2.5 (EUR)   
20                                                              Шестигранный HD Болт, 7/16" Х 1-3/4"L, GR8   
21     Ноутбук Latitude 5520 Core i5-1135G7 2.4GHz 15,6" FullHD IPS Antiglare 400 nits 16GB DDR4 512GB SSD   
22                                                     Карта сухих контактов Relay Card-MS бля ИБП, Eaton,   
23                                           Клапан трехходовой термостатический системы охлаждения -110°F   
24                                                    Механический байпас для вывода ИБП из работы, Eaton,   
25                                                                            4-ХОД. КЛАП. КОРОБКА В СБОРЕ   
26                                                                                    4-ХОД КЛАПАН В СБОРЕ   
27                                                                       Стык-пакет Super Screw SS-80 A NO   
28                                                                                 Trellex PP  40-500-1500   
29                                                                    Плита футеровочная 40/5-600-1500-T60   
30                                                                                 Trellex WB 100-100-1500   
31                                                                                        Прижимная полоса   
32                                                                                 Угловая прижиная полоса   
33                                                                                Угловая прижимная полоса   
34                                                            Сетевой фильтр PILOT-S (6/5м/10А/80Дж/белый)   
35                                                                                     GEAR BOX TURRET DRV   
36                      

In [17]:
n0 = 0
n1 = 0
n2 = 0
for i in noun_list:
    if len(i) == 1:
        n1 +=1
    elif len(i) > 1:
        n2 +=1
    elif not len(i):
        n0 +=1
print(f"существительные определены в {n1+n2} строках \nв том числе более одного существительного в {n2} строках")

существительные определены в 882 строках 
в том числе более одного существительного в 196 строках


### Итог
1. Для большинства строк (882) были определены существительные в столбце "существительные", для строк на латинице существительные не определялись.
2. При определении существительных в строке наименования приоритет был отдан существительным в именительном падеже, так как они в большей степени отражают сущность наименования. Существительные в других падежах и или во множественном числе могли быть определены не во всех случаях из за особенностей словообразования.
3. В отдельных случаях прилагательные (белый, цветной) могли быть определены как существительные
4. Для более точного определения частей речи в наборах данных большего объема можно использовать словарь большего объема и учетом различных форм слов (корпус современного русского языка) а так же методы ML.


## Поиск дубликатов
Проводится подготовка данных аналогичная той, что использовалась для поиска существительных. Заменяются сокращения по ранее созданному словарю

In [19]:
names_dub = df['Наименование'].tolist()

# перевод символов в нижний регистр
for i in range(len(names_dub)):
    names_dub[i] = names_dub[i].lower()

# замена символов на пробел
for i in range(len(names_dub)):
    names_dub[i] = replace_all('[-)/,(]', names_dub[i], ' ')

# вставка пробела после точек и запятых
for i in range(len(names_dub)):
    names_dub[i] = names_dub[i].replace('.', '. ', names_dub[i].count('.')).replace(',', ', ', names_dub[i].count(','))

# удаление лишних пробелов 
for i in range(len(names_dub[i])):
    names_dub[i] = clean_spaces(names_dub[i])

In [20]:
# замена сокращений
for i in range(len(names_dub)):
    line = names_dub[i].split()
    for j in line:
        for word in dict_shorts:
            if j.startswith(word):
                names_dub[i] = names_dub[i].replace(word, dict_shorts[word])

In [28]:
# перечень потенциальных дубликатов
names_dubl = pd.Series(names_dub)
df['names dubl'] = names_dubl.values
df[df.duplicated(['names dubl'], keep=False)].sort_values(by='names dubl')

Наименование  \
221                                                                             FLANGE PIPE   
138                                                                             FLANGE PIPE   
105  Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
106  Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
107  Injection moulded PU trommel panel 305x305x36mm Thick/12x50mm apertures, Hi-temp/150mm   
740                                                           PROTECTION PLATE INSTALLATION   
619                                                           PROTECTION PLATE INSTALLATION   
193                                                                                  PULLER   
127                                                                                  PULLER   
548                                                                                   SHAFT   
683                                                                                   SHAFT   
324                                                                          SPRING SUPPORT   
325                                                                          SPRING SUPPORT   
302                                                                          SUPPORT HOLDER   
531                                                                          SUPPORT HOLDER   
826                                              TRELLEX BAR-COMP. T-BOLT TF35/25 M12x40x35   
825                                              TRELLEX BAR-COMP. T-BOLT TF35/25 M12x40x35   
793                                                          TRELLEX IMPACT BAR 75x100x1220   
794                                                          TRELLEX IMPACT BAR 75x100x1220   
809                                                              Trellex LF S9 15-1220-3050   
808                                                              Trellex LF S9 15-1220-3050   
830                                                            Trellex LF S9 20-1220-3050 W   
829                                                            Trellex LF S9 20-1220-3050 W   
804                                                               TRELLEX LF S9 6-1220-3015   
805                                                               TRELLEX LF S9 6-1220-3015   
353                                                                  Trellex PP 75-500-1000   
354                                                                  Trellex PP 75-500-1000   
689                                                                              WEAR PLATE   
695                                                                              WEAR PLATE   
690                                                                              WEAR PLATE   
338                                                                              WEAR PLATE   
706                                                                              WEAR PLATE   
558                                                                                   WEDGE   
557                                                                                   WEDGE   
232                                                                                   WEDGE   
671                                                                                   WEDGE   
144                                                       Аксессуар SMS-L ACC Nut M16 nyloc   
143                                                       Аксессуар SMS-L ACC Nut M16 nyloc   
129                                                 БОЛТ С ШЕСТИГР.ГОЛ ISO4014-M16X40-8.8-A   
130                                                 БОЛТ С ШЕСТИГР.ГОЛ ISO4014-M16X40-8.8-A   
335                                                                          БРОНЯ ДРОБИЛКИ   
334                                                                          БРОНЯ ДРОБИЛ